smart_open/tests/test_s3_version.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+import logging
+import os
+import time
+import unittest
+import uuid
+
+import boto3
+import botocore.client
+import moto
+
+import smart_open
+from smart_open import open
+
+BUCKET_NAME = 'test-smartopen-{}'.format(uuid.uuid4().hex)
+KEY_NAME = 'test-key'
+WRITE_KEY_NAME = 'test-write-key'
+DISABLE_MOCKS = os.environ.get('SO_DISABLE_MOCKS') == "1"
+
+
+logger = logging.getLogger(__name__)
+
+
+def maybe_mock_s3(func):
+    if DISABLE_MOCKS:
+        return func
+    else:
+        return moto.mock_s3(func)
+
+
+@maybe_mock_s3
+def setUpModule():
+    '''Called once by unittest when initializing this module.  Sets up the
+    test S3 bucket.
+
+    '''
+    boto3.resource('s3').create_bucket(Bucket=BUCKET_NAME)
+    boto3.resource('s3').BucketVersioning(BUCKET_NAME).enable()
+
+
+def put_to_bucket(contents, num_attempts=12, sleep_time=5):
+    # fake (or not) connection, bucket and key
+    logger.debug('%r', locals())
+
+    #
+    # In real life, it can take a few seconds for the bucket to become ready.
+    # If we try to write to the key while the bucket while it isn't ready, we
+    # will get a ClientError: NoSuchBucket.
+    #
+    for attempt in range(num_attempts):
+        try:
+            boto3.resource('s3').Object(BUCKET_NAME, KEY_NAME).put(Body=contents)
+            return
+        except botocore.exceptions.ClientError as err:
+            logger.error('caught %r, retrying', err)
+            time.sleep(sleep_time)
+
+    assert False, 'failed to create bucket %s after %d attempts' % (BUCKET_NAME, num_attempts)
+
+
+@maybe_mock_s3
+class TestVersionId(unittest.TestCase):
+    def setUp(self):
```

Comment:
This method gets run each time you run the test. So, it gets run around 4 times. Each time you run the test, you're adding two new versions of the file, so by the end of the entire test suite, you have **eight** versions of the file. This is bad, because it means your tests have side effects, and they can interfere with each other.

I think you should move this version creation logic to the top-level code that creates the bucket, so those versions are only created once.